src/main/java/io/supertokens/storage/postgresql/queries/UserRolesQueries.java

```
@@ -48,74 +56,117 @@ public static String getQueryToCreateRolePermissionsTable(Start start) {
+                + ");";
         // @formatter:on
     }
 
     public static String getQueryToCreateUserRolesRoleIndex(Start start) {
-        return "CREATE INDEX user_roles_role_index ON " + getConfig(start).getUserRolesTable() + "(role);";
+        return "CREATE INDEX user_roles_role_index ON " + getConfig(start).getUserRolesTable()
+                + "(app_id, tenant_id, role);";
     }
 
-    public static boolean createNewRoleOrDoNothingIfExists_Transaction(Start start, Connection con, String role)
-            throws SQLException, StorageQueryException {
-        String QUERY = "INSERT INTO " + getConfig(start).getRolesTable() + " VALUES(?) ON CONFLICT DO NOTHING;";
-        int rowsUpdated = update(con, QUERY, pst -> pst.setString(1, role));
-        return rowsUpdated > 0;
+    public static boolean createNewRole_Transaction(Start start, Connection con,
+                                                    AppIdentifier appIdentifier, String role)
+            throws SQLException, StorageQueryException, TenantOrAppNotFoundException {
+        {
+            // Checking if app_id exists because in the next query conflicts are ignored
+            String QUERY = "SELECT app_id FROM " + getConfig(start).getAppsTable()
+                    + " WHERE app_id = ?;";
+            String returnedAppId = execute(con, QUERY, pst -> {
+                pst.setString(1, appIdentifier.getAppId());
+            }, result -> {
+                if (result.next()) {
+                    return result.getString("app_id");
+                }
+                return null;
+            });
+            if (returnedAppId == null) {
+                throw new TenantOrAppNotFoundException(appIdentifier);
+            }
+        }
+        {
+            String QUERY = "INSERT INTO " + getConfig(start).getRolesTable()
+                    + "(app_id, role) VALUES (?, ?) ON CONFLICT DO NOTHING;";
```

add a comment saying that "ON CONFLICT DO NOTHING" will ignore errors if appId is not actually valid. But we are OK with that cause its an edge case wherein the app is deleted right before this query is happening, and we are ok with that.

---

Not required, it was mistake in my understanding. Foreign key constraint errors are still thrown. Verified that on my postgres db.